src/frontend/parser.mly

```diff
@@ -218,61 +232,125 @@ unsized_dims:
   | LBRACK cs=list(COMMA) RBRACK
     { grammar_logger "unsized_dims" ; List.length(cs) }
 
-(* declarations *)
+(* Never accept this rule, but return the same type as expression *)
+no_assign:
+  | UNREACHABLE
+    { (* This code will never be reached *)
+      raise (Failure "This should be unreachable; the UNREACHABLE token should \
+                      never be produced")
+    }
+
+optional_assignment(rhs):
+  | rhs_opt=option(pair(ASSIGN, rhs))
+    { Option.map ~f:snd rhs_opt }
+
+id_and_optional_assignment(rhs):
+  | id=decl_identifier rhs_opt=optional_assignment(rhs)
+    { (id, rhs_opt) }
+
+(*
+ * All rules for declaration statements.
+ * The first argument matches the type and should return a (type, constraint) pair.
+ * The second argument matches the RHS expression and should return an expression
+ *   (or use no_assign to never allow a RHS).
+ *
+ * The value returned is a function from a bool (is_global, which controls
+ * whether the declarations should be global variables) to a list of statements
+ * (which will always be declarations).
+ *
+ * The rules match declarations with/without assignments, with/without array
+ * dimensions, single/multiple identifiers, and dimensions before/after the
+ * identifier.
+ *)
+decl(type_rule, rhs):
+  (* When dims are after identifier, do not allow multiple identifiers *)
+  | ty=type_rule id=decl_identifier dims=dims rhs_opt=optional_assignment(rhs)
```

Comment:
So this is matching the new syntax and the next production is matching the old syntax, right? I would love a comment that says that and just gives a concrete example - would help me reading this code again later. Could also mark the old syntax with some "TO BE DEPRECATED" string we can then search from when Stan 3.0 does come out and break things.